### DISCUSSION TITLE
How to implement UnitOfWork on Csla 8

### DISCUSSION THREAD

**Discussion by ecaldentey** [MarimerLLC/csla#4108] [Questions]
Hi Rocky,

Could you guide me on how to implement a UnitOfWork in CSLA 8?

I need to perform a series of operations on different CSLA objects within a single transaction, such as editing some records in a BusinessListBase, updating a root BusinessBase with children, etc. In other words, multiple operations across different classes, and if any operation fails, revert everything.

Currently, I can do this separately, one class at a time, but if something fails, I have to delete what has been done class by class, which is insecure and inefficient.

In fact, the updates of the BusinessListBase are a kind of UnitOfWork since this class opens the database connection, starts the transaction, then passes control to the child classes that are root BusinessBase with children. But as they know they are part of a BusinessListBase, the items in the list that are root objects have their IsChild property set to true. They skip opening the database connection because they know the BusinessListBase did it, and the same with the transaction. If something happens with any object in the list, the BusinessListBase takes care of the rollback, and if everything is fine, the BusinessListBase commits and informs the UI. In my opinion, it is a good example of UnitOfWork.

What I want to do is something similar but with a UnitOfWork object that contains all the objects involved in the transaction and that is responsible for making the database connection and opening the transaction. If any of the different objects contained in the UnitOfWork report an error, the UnitOfWork should handle the rollback and report the error to the UI; otherwise, it should commit.

Can you guide me on how to do this in CSLA 8 or indicate where I can find an example? I have reviewed the CSLA projects and forums but haven't found clear information regarding what I need to do.

Thank you

**Comment by rockfordlhotka** [comment]
What I am hearing, is that you have multiple object graphs, each with its own root object, and you need to update these object graphs within a single unit of work?

The easiest way to implement the UoW pattern is typically as a command object that inherits from `CommandBase`. Here is some pseudocode:

```c#
public class MyUnitOfWork : CommandBase<MyUnitOfWork>
{
  // implement one property for each root object

  public static readonly PropertyInfo<MyRoot1> MyRoot1Property = RegisterProperty<MyRoot1>(nameof(MyRoot1));
  public MyRoot1 MyRoot1
  {
    get => ReadProperty(MyRoot1Property);
    set => LoadProperty(MyRoot1Property, value);
  }

  [Create]
  [RunLoca]
  private void Create()
  {}

  [Execute]
  private async Task Execute([Inject] ISqlTransaction transaction, [Inject] IDataPortal<MyRoot1>)
  {
    transaction.Start()
    MyRoot1 = await MyRoot1.SaveAsync();
    // repeat for other object graphs
    transaction.Commit();
  }
}

// calling code with an IDataPortal<MyUnitOfWork> instance
var uow = await commandPortal.CreateAsync();
uow.MyRoot1 = myRoot1;
// set other object graph properties
uow = await commandPortal.ExecuteAsync(uow);
// you now have all the updated object graphs
```

**Reply by ecaldentey** [reply]
Hi Rocky,

Thank you for your time and the comprehensive response.

What I need is a single trip to the Dataportal to move several completely "independent" objects with their graphs, open a single connection to the database, update all the objects with their respective graphs, and once the last update is done, commit the transaction or rollback if there was an error, and close the database connection.

public class MyUnitOfWork : CommandBase<MyUnitOfWork>
{
    // implement one property for each root object

    public static readonly PropertyInfo<MyRoot1> MyRoot1Property = RegisterProperty<MyRoot1>(nameof(MyRoot1));
    public MyRoot1 MyRoot1
    {
        get => ReadProperty(MyRoot1Property);
        set => LoadProperty(MyRoot1Property, value);
    }
  
    public static readonly PropertyInfo<MyListBase> MyListBaseProperty = RegisterProperty<MyRoot1>(nameof(MyListBase));
    public MyListBase MyListBase
    {
        get => ReadProperty(MyListBaseProperty);
        set => LoadProperty(MyListBaseProperty, value);
    }
....

When I execute the following instruction:
uow = await commandPortal.ExecuteAsync(uow);

At that moment, CSLA handles moving the objects to the dataportal method Execute of the command object and then when executing the following code:
MyRoot1 = MyRoot1.SaveAsync();

CSLA moves to the Update dataportal method of my root object and if I have:

MyListBase = MyListBase.SaveAsync();
in this case, CSLA moves to the Update dataportal method of my BusinessListBase object.

I handle everything from the dataportal Execute method and commit or rollback and close the database from that method.

Am I understanding correctly?

Thanks in advance

**Reply by rockfordlhotka** [reply]
Yes, that is correct.

**Reply by ecaldentey** [reply]
Hi Rocky,

I have a question regarding using a CommandBase to construct a UnitOfWork. The properties of the command object will be BusinessBase root or BusinessListBase which by design are not "Child" objects and when invoking the Save method, by design, they open a database connection and a transaction.

In the normal case of a BusinessListBase, for example, its items, which are BusinessBase root, do not open a database connection and transaction because by class design, their property is marked as IsChild = true. Therefore, in the Dataportal method, this property is checked, and if the root class is marked as IsChild, it knows that it is a "child" of a list, thus it "knows" that the connection and transaction are already established.

So when I create a property in a CommandBase object that is a BusinessBase root type and, for example, another property that is a BusinessListBase, which in this case is a "child" or something like that of the CommandBase object, how do I mark this condition so that in the Dataportal methods of both the BusinessBase root object and the BusinessListBase, it can read this condition and skip opening the database connection and transaction because they are already handled by the CommandBase object?

How do I do the above? How do I mark them as "IsChild" or something similar? Does CSLA have a property for these purposes or should I modify my class design to be able to pass a property that allows identifying them as properties of a CommandBase object and part of a UnitOfWork?

Thanks for the help.

**Reply by rockfordlhotka** [reply]
Command objects are different from `BusinessBase` or `BusinessListBase`. A command object isn't a "parent" - it doesn't have that concept.

What we have been discussing is that you need a unit of work that operates on a set of _root_ objects within a single transaction.

A _root_ object might be a single object, or the root of an _object graph_, where all the other objects in the graph are children (or grandchildren) of the root object.

I am assuming you've read the free CSLA intro book from https://store.lhotka.net? Perhaps the "Using CSLA 4" books too?

**Reply by ecaldentey** [reply]
Hi Rocky,

I think I didn't explain myself well, and I apologize for that. I'll try to clarify my question.

Of course, I've read all your books, and I understand very well that the CommandBase object works very differently from a BusinessBase or BusinessListBase object, or their ReadOnly variants, etc.

I understand clearly that a property of a CommandBase object is not a Child object as understood in the CSLA ecosystem, and its type could be any type of object, a string, an integer, a serializable object, or a CSLA class of any nature as is the case in what we are discussing.

What we are discussing is the possibility of using a CommandBase object to construct a UnitOfWork, where we are taking advantage of the CommandBase object's ability to define multiple properties and also move them to the Dataportal side using all the capabilities of CSLA in its Execute method, and I understood that perfectly.

Obviously, a CommandBase object does not have the concept of having children; it has properties. That is not under discussion. My problem is how to indicate to the properties of the CommandBase object, which in our specific case are of type CSLA, to behave as if they were "children" or if you prefer "subordinates" of the CommandBase object so that when the CommandBase object moves them to the Dataportal, they do not perform actions they have defined by design, such as opening the database connection and starting the transaction. Do you understand what I mean?

For example, by design, a BusinessBase object that is a root, when doing a Save, moves the entire graph (its children) to the Dataportal and opens a database connection, using a Scoped service for that, and the same with the transaction. It shares with its children, which are BusinessBase classes but child objects, the database connection and transaction from its parent object, and they perform their actions. If these have children, they cascade the database connection and transaction downwards. When the entire graph is updated, everything returns to the parent or root object, and it is responsible for committing or rolling back the transaction and closing the database connection.

But in our case, using a CommandBase as a controller or "parent" of the UnitOfWork somehow needs to tell its properties, in this case of type CSLA BusinessBase root, BusinessBaseList, etc., not to open database connections or transactions because it is responsible, and it is the CommandBase that will commit or rollback and close the database connection.

That is my question. How does the CommandBase tell its properties that it controls everything, and the properties behave as if they were children or subordinates to the CommandBase?

I hope I've explained it better.

Thank you.

**Reply by rockfordlhotka** [reply]
That helps, thank you.

What you are looking for is the concept of composability when it comes to data access, where a root object can be composed with other root objects into a UoW, such that it reuses any pre-existing database transaction.

The easiest way to do this is to use the `Csla.Transactional` attribute with the `TransactionalTypes.TransactionScope` option. This uses the built-in ADO.NET transaction manager. As long as all your objects use the same database connection, they should be enlisted in the same transaction scope.

If you are injecting the connection object as a DI service into your DAL objects (which you should be), and the connection object is a _scoped_ service, then the connection should be reused across all operations in the unit of work, and should be enlisted in a transaction thanks to ADO.NET.

```c#
        [Update]
        [Transactional(TransactionalTypes.TransactionScope)]
        private async Task UpdateAsync([Inject] PersonDal dal)
        {
            await dal.UpdateAsync(new PersonEntity { Name = Name });
        }
```

and

```c#
        [Execute]
        [Transactional(TransactionalTypes.TransactionScope)]
        private async Task ExecuteAsync([Inject] IDataPortal<PersonEdit> personPortal)
        {
            await personPortal.UpdateAsync(Person);
        }
```

**Reply by ecaldentey** [reply]
Indeed, I do as you suggest and use DI to share the database connection service, injecting it into the Dataportal methods because my object interfaces have the definitions.

In the case of a root BusinessBase object, in the Dataportal method, I check if it is a child (IsChild Csla property this.IsChild = true) and with that I know it is a child of a BusinessListBase. With this condition, I perform actions in the database and then call: this.FieldManager.UpdateChildren(this);

On the other hand, if this.IsChild = false, in the Update method of the BusinessClass, I open the database connection and start the transaction.

Therefore, when the root BusinessBase is a property of a CommandBase, I don't know how to recognize this condition. I lack a property, and I am not sure if I need to modify the design of my classes to add a property that allows me to identify the condition of being a property of a CommandBase object or if I can use some Csla property.

Here is an excerpt of how my interfaces are and how the Dataportal methods of a root BusinessBase class look, to better explain the problem. For this example, I simulate the existence of a property IsCommandProperty.

public interface ICustomerDal
{
    #region [Interface_Methods]

	DbTransaction StartTransaction(Int64 transactionId = 0);
	void EndTransaction(bool commit, Int64 transactionId = 0);
	void CloseConnection();
	decimal FetchId(CustomerGetIdDbTypes getIdDbType = CustomerGetIdDbTypes.CustomerId, string customFilters = "");
	CustomerDto FetchObject(CustomerDto customerDto, CustomerSelectDbTypes selectDbType, string customFilters);
	List<CustomerDto> FetchList(CustomerDto customerDto, CustomerSelectDbTypes selectDbType, string customFilters);
	ResultMethod InsertObject(CustomerDto customerDto, CustomerInsertDbTypes insertDbType, string customFilters);
	ResultMethod UpdateObject(CustomerDto customerDto, CustomerUpdateDbTypes updateDbType, string customFilters);
	ResultMethod DeleteObject(CustomerDto customerDto, CustomerDeleteDbTypes deleteDbType, string customFilters);

	#endregion [Interface_Methods]
}

//Dataportal Methods
#region [Insert_Update_Delete]

[Insert]
[InsertChild]
private void Insert([Inject] ICustomerDal dal, [Inject] ILanguageService serviceLanguage)
{
	PerformUpdate(() => dal.InsertObject(LoadData(), (CustomerInsertDbTypes)(this.OptionSelector ?? CustomerInsertDbTypes.InsertDefault), this.CustomFilters), dal);
}

[Update]
[UpdateChild]
private void Update([Inject] ICustomerDal dal, [Inject] ILanguageService serviceLanguage)
{
	PerformUpdate(() => dal.UpdateObject(LoadData(), (CustomerUpdateDbTypes)(this.OptionSelector ?? CustomerUpdateDbTypes.UpdateDefault), this.CustomFilters), dal);
}

[Delete]
[DeleteSelfChild]
private void Delete([Inject] ICustomerDal dal, [Inject] ILanguageService serviceLanguage)
{
	PerformUpdate(() => dal.DeleteObject(LoadData(), (CustomerDeleteDbTypes)(this.OptionSelector ?? CustomerDeleteDbTypes.DeleteDefault), this.CustomFilters), dal);
}

private void PerformUpdate(Func<ResultMethod> databaseOperation, ICustomerDal dal)
{
  
	if (this.IsChild || this.IsCommandProperty)
	{
		//!!!Code when the BusinessBase root is child of a BusinessListBase or property of a CommandBase!!!
		DoUpdate(databaseOperation, out bool resultStatus, out string resultMsg);

		if (!resultStatus)
		{
			throw new Exception(resultMsg);
		}
	}
	else
	{
		//!!!Code when the root BusinessBase object is operating on its own 
		//and directly controlling the database connection and transaction !!!
		using (var conn = dal.StartTransaction())
		{
			#region [StartTransaction]

			DoUpdate(databaseOperation, out bool resultStatus, out string resultMsg);

			dal.EndTransaction(resultStatus);

			if (!resultStatus)
			{
				throw new Exception(resultMsg);
			}

			#endregion [StartTransaction]
		}
	}
}

private void DoUpdate(Func<ResultMethod> databaseOperation, out bool resultStatus, out string resultMsg)
{
	resultStatus = true;
	resultMsg = String.Empty;

	var resultMethod = databaseOperation();

	resultStatus = resultMethod.ResultStatus;

	if (resultStatus)
	{
		try
		{
			this.FieldManager.UpdateChildren(this);
		}
		catch (Exception ex)
		{
			resultStatus = false;
			resultMsg = ex.Message;
		}
	}
	else
	{
		resultMsg = resultMethod.ResultMsg;
	}
}

private CustomerDto LoadData()
{
	CustomerDto customerDto = new CustomerDto();

	customerDto.CustomerId = this.CustomerId;
	customerDto.CompanyName = this.CompanyName;
	....
	/*----------- Old Properties -----------*/
	customerDto.OldCustomerId = this.OldCustomerId;

	return customerDto;
}

#endregion [Insert_Update_Delete]

**Reply by rockfordlhotka** [reply]
> On the other hand, if this.IsChild = false, in the Update method of the BusinessClass, I open the database connection and start the transaction.

If you inject an open connection and use the `Transactional` attribute, then you don't need to open connections or transactions (or commit/rollback) in your code. This should make your code composable, because your root objects shouldn't worry about whether the connection it got injected was already associated with a transaction.

**Reply by rockfordlhotka** [reply]
@ecaldentey here's a gist that shows the use of the `Transactional` attribute and an injected connection object. It isn't a unit of work scenario, but it does demonstrate how to allow DI to open the connection, and the data portal to manage the transaction. This way none of your DAL code worries about those details.

https://gist.github.com/rockfordlhotka/e3f6f1cd5e030cfc474818a3f418095d

**Reply by ecaldentey** [reply]
Hello Rocky,

Thank you for the comprehensive response, and for your time and dedication.

I have currently developed a code generator based on Codesmith that generates all the classes. What I am going to do is remove the references from my interfaces to the methods that handle transactions and add the attribute [Transactional(TransactionalTypes.TransactionScope)] to the dataportal methods. This will allow me to avoid managing transactions manually, which should greatly simplify my code.

Thank you very much for the help.
